pkg/operator/ceph/cluster/pdb.go

```diff
@@ -0,0 +1,87 @@
+package cluster
+
+import (
+	"reflect"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const drainDetectionLabel = "app=rook-ceph-drain-detection"
+
+func (c *cluster) startPDBController() {
```

Comment:
Looking through the ceph operator code, I'm not seeing any workqueues and reconcile loops with rate limiting. All the reconcile logic is apparently being done within informer callbacks?! 

This isn't the operator pattern. Informer callbacks need to be pushing keys onto a work queue, and the reconcile loop should be popping items off the work queue in a separate go-routine.

Without using the workqueue pattern, controllers like this are not resilient to failures.  For example, if the reconcile logic that is invoked during the informer callback happens to fail for whatever reason (api server unavailable, network disconnects, etc...) then it will never be ratelimited and retried later on. 

With a workqueue we have the ability to re-enqueue a key to re-execute later on in the event that an error is encountered. Executing directly from the informer callbacks does not give us that ability.

Reply:
here's an example of what I'm talking about taken directly from the k8s deployment controller [source](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L148).

```
// Run begins watching and syncing.
func (dc *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
	defer utilruntime.HandleCrash()
	defer dc.queue.ShutDown()

	klog.Infof("Starting deployment controller")
	defer klog.Infof("Shutting down deployment controller")

	if !controller.WaitForCacheSync("deployment", stopCh, dc.dListerSynced, dc.rsListerSynced, dc.podListerSynced) {
		return
	}

	for i := 0; i < workers; i++ {
		go wait.Until(dc.worker, time.Second, stopCh)
	}

	<-stopCh
}
```

The controller loop starts by ensuring that all the informer's that it is watching are synced and up to date, then the controller loop starts a fixed number of worker threads to process the work queue.

The informer callbacks push keys onto the queue that represent deployments, then the workqueue goroutines pop those keys off and process the deployment's reconcile loop. They aren't executing the actual reconcile loop logic within the informer callbacks though, because that isn't resilient to failure (no retries) and it doesn't make the concurrency guarantee that only a single deployment is acted on at a time.

Reply:
Yeah, I did notice that we were not following the traditional controller pattern. I didn't have time to figure out the pattern from reading code, but I did try by reading controller-runtime code. I outlined some issues in Issue #3577. One easy way to get the controller pattern would be to use controller-runtime, but I didn't understand it enough yet to push for it in rook.

Other parts of rook do implement retries, and I can implement them here. I can also process each reconciled object in a dedicated goroutine, but that still misses a lot of stuff.

What I like better is to import controller-runtime and implement a manager and a controller which provides all this, but I don't want to make this PR about replacing all the controllers with controller-runtime as that will take too much time.

@travisn @davidvossel could we have a call about this?

Reply:
I have since read [kubernetes/sample-controller](https://github.com/kubernetes/sample-controller) I see that queues, rate limiting, etc are simple enough to implement here.

Reply:
> What I like better is to import controller-runtime and implement a manager and a controller which provides all this, but I don't want to make this PR about replacing all the controllers with controller-runtime as that will take too much time

Yeah... I know what you mean. This is a tough situation.

It's unfortunate, but we've got to fix this controller situation. Realistically, the pattern being replicating here isn't supportable. Even with retries, we'll likely encounter situations where the logic simply stops working. I don't believe the informer callbacks aren't designed to work this way and be blocked like this. Adding retries in attempt to make the logic more resilient actually compounds other issues involved with concurrency.

I agree with what you're saying here though. This shouldn't be the topic of this specific PR. At the same time, I don't see a way for this PR to proceed without fixing the broken pattern we're using.

I think revising the controller logic flow is something we should break out into a separate PR. I'll try and invest some time into understanding how large of a task this is and how best to approach the issue. This might need to be a joint effort.

Reply:
@davidvossel take a look at the latest changes. I've implemented a scaffolding for controller-runtime that I can use for this controller.

It also establishes a pattern that other controllers can use to  move into controller-runtime in their own PRs.